### PR TITLE
Jr pundit

### PR DESCRIPTION
**Pull request by jb520**
Created policies for User, Photo, Comment, Likes, and Follow Requests.

### CODE REVIEW

**Review by MayaS1111 — APPROVED** [review]
Looks great! Keep up the good work!

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,12 +1,27 @@
 class ApplicationController < ActionController::Base
+  include Pundit
+
   before_action :authenticate_user!
 
   before_action :configure_permitted_parameters, if: :devise_controller?
 
+  after_action :verify_authorized, unless: :devise_controller?
```
Comment (ajmompr) [inline app/controllers/application_controller.rb:8]: good job accounting for devise

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,12 +1,27 @@
...
+  after_action :verify_policy_scoped, only: :index, unless: :devise_controller?
```
Comment (ajmompr) [inline app/controllers/application_controller.rb:9]: good job accounting for devise

File: app/controllers/comments_controller.rb [hunk 3]
```diff
@@ -1,6 +1,7 @@
 class CommentsController < ApplicationController
   before_action :set_comment, only: %i[ show edit update destroy ]
   before_action :is_an_authorized_user, only: [:destroy, :create]
+  before_action{ authorize(@comment || Comment)}
```
Comment (ajmompr) [inline app/controllers/comments_controller.rb:4]: good!

File: app/controllers/follow_requests_controller.rb [hunk 4]
```diff
@@ -1,6 +1,6 @@
 class FollowRequestsController < ApplicationController
   before_action :set_follow_request, only: %i[ show edit update destroy ]
-
+  before_action{ authorize(@follow_request || FollowRequest)}
```
Comment (ajmompr) [inline app/controllers/follow_requests_controller.rb:3]: good!

File: app/controllers/likes_controller.rb
Comment (ajmompr) [inline app/controllers/likes_controller.rb:1]: we don't need to add pundits to likes for this assignment!

File: app/controllers/photos_controller.rb [hunk 5]
```diff
@@ -1,14 +1,16 @@
 class PhotosController < ApplicationController
   before_action :set_photo, only: %i[ show edit update destroy ]
   before_action :ensure_current_user_is_owner, only: [:destroy, :update, :edit]
-
+  before_action :ensure_user_is_authorized, only: [:show]
```
Comment (ajmompr) [inline app/controllers/photos_controller.rb]: do we still need this if we are defining our policy and authorizing in the before_action below?

File: app/controllers/photos_controller.rb [hunk 6]
```diff
@@ -1,14 +1,16 @@
...
+  before_action{ authorize(@photo || Photo)}
   # GET /photos or /photos.json
   def index
     @photos = Photo.all
   end
 
   # GET /photos/1 or /photos/1.json
   def show
+    authorize @photo
```
Comment (ajmompr) [inline app/controllers/photos_controller.rb]: if you are calling the before_action { authorize ... }, this is redundant

File: app/controllers/photos_controller.rb [hunk 7]
```diff
@@ -70,6 +72,12 @@ def ensure_current_user_is_owner
       end
     end
 
+    def ensure_user_is_authorized
```
Comment (ajmompr) [inline app/controllers/photos_controller.rb]: do we stll need this?

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -1,5 +1,6 @@
 class UsersController < ApplicationController
   before_action :set_user, only: %i[ show liked feed followers following discover ]
+  before_action{ authorize(@user || User)}
```
Comment (ajmompr) [inline app/controllers/users_controller.rb:3]: good!

File: app/policies/application_policy.rb [hunk 9]
```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+class ApplicationPolicy
```
Comment (ajmompr) [inline app/policies/application_policy.rb:3]: good use of pundit generator

File: app/policies/comment_policy.rb
Comment (ajmompr) [inline app/policies/comment_policy.rb:1]: Great job! There is one little edge case to account for here, but I will not take points off.

File: app/policies/follow_request_policy.rb
Comment (ajmompr) [inline app/policies/follow_request_policy.rb:1]: perfect

File: app/policies/like_policy.rb
Comment (ajmompr) [inline app/policies/like_policy.rb:1]: Not needed for this assignment; likes are not fully implemented

File: app/policies/photo_policy.rb
Comment (ajmompr) [inline app/policies/photo_policy.rb:1]: good

File: app/policies/user_policy.rb
Comment (ajmompr) [inline app/policies/user_policy.rb:1]: Looks good!

**Review by ajmompr — CHANGES_REQUESTED** [review]

File: app/controllers/likes_controller.rb
Comment (jb520) [reply app/controllers/likes_controller.rb:1]: removed likes policy!

File: app/controllers/photos_controller.rb (hunk 5 above)
Comment (jb520) [reply app/controllers/photos_controller.rb]: refactored photos controller!

File: app/controllers/photos_controller.rb (hunk 6 above)
Comment (jb520) [reply app/controllers/photos_controller.rb]: refactored photos controller!

File: app/controllers/photos_controller.rb (hunk 7 above)
Comment (jb520) [reply app/controllers/photos_controller.rb]: refactored photos controller!

File: app/policies/like_policy.rb
Comment (jb520) [reply app/policies/like_policy.rb:1]: removed likes policy!

File: app/views/users/show.html.erb [hunk 10]
```diff
@@ -4,7 +4,7 @@
   </div>
 </div>
 
-<% if current_user == @user || @user.private? || current_user.leaders.include?(@user) %>
+<% if policy(@user).show? %>
```
Comment (ghost) [inline app/views/users/show.html.erb]: Oh, I think I didn't call this policy here. I believe I did somewhere else, but maybe I should go back and adjust that quick!

**Review by ghost — APPROVED** [review]
Looks good to me!

File: app/controllers/likes_controller.rb
Comment (ajmompr) [inline app/controllers/likes_controller.rb:1]: no need to add pundit to likes for this assignment

File: app/controllers/photos_controller.rb [hunk 11]
```diff
@@ -1,7 +1,7 @@
 class PhotosController < ApplicationController
   before_action :set_photo, only: %i[ show edit update destroy ]
   before_action :ensure_current_user_is_owner, only: [:destroy, :update, :edit]
-
+  before_action{ authorize(@photo || Photo)}
```
Comment (ajmompr) [inline app/controllers/photos_controller.rb:4]: good

File: app/policies/follow_request_policy.rb
Comment (ajmompr) [inline app/policies/follow_request_policy.rb:1]: good

File: app/policies/photo_policy.rb
Comment (ajmompr) [inline app/policies/photo_policy.rb:1]: good

File: app/policies/user_policy.rb [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+  def initialize(current_user, user)
+    @current_user = current_user
+    @user = user
+  end
+
+  def feed?
+    true
+  end
+
+  def show?
```
Comment (ajmompr) [inline app/policies/user_policy.rb:13]: almost! Is your app displaying properly for users who are private. Try moving this logic to a method called show_photos? and setting show to true. Call show_photos? policy in your users#show template

**Review by ajmompr — CHANGES_REQUESTED** [review]
One more change and you are good. See comments on user policy and DM me once fixed.

File: app/policies/user_policy.rb (hunk 12 above)
Comment (jb520) [reply app/policies/user_policy.rb:13]: show_policy created!

File: app/policies/user_policy.rb [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+  def destroy?
+    update?
+  end
+
+  def discover?
+    update?
+  end
+
+  def show_photos?
```
Comment (ajmompr) [inline app/policies/user_policy.rb:39]: good

File: app/views/users/show.html.erb [hunk 14]
```diff
@@ -4,7 +4,7 @@
   </div>
 </div>
 
-<% if current_user == @user || @user.private? || current_user.leaders.include?(@user) %>
+<% if policy(@user).show_photos? %>
```
Comment (ajmompr) [inline app/views/users/show.html.erb:7]: perfect

**Review by ajmompr — APPROVED** [review]
perfect

**Pull request merged**